concrete/testtool/testtool.go

```diff
@@ -43,84 +42,81 @@ var (
+		senderAddress := crypto.PubkeyToAddress(key.PublicKey)
+		contractAddress := common.HexToAddress("cc73570000000000000000000000000000000000")
+		gspec := &core.Genesis{
 			GasLimit: 2e9,
 			Config:   params.TestChainConfig,
 			Alloc: types.GenesisAlloc{
 				senderAddress:   {Balance: big.NewInt(1e18)},
 				contractAddress: {Balance: common.Big0, Code: bytecode},
 			},
 		}
-		signer   = types.LatestSigner(gspec.Config)
-		gasLimit = uint64(1e7)
-		setupId  = crypto.Keccak256([]byte("setUp()"))[:4]
-	)
+		signer := types.LatestSigner(gspec.Config)
+		gasLimit := uint64(1e7)
+		setupId := crypto.Keccak256([]byte("setUp()"))[:4]
 
-	_, _, receipts := core.GenerateChainWithGenesisWithConcrete(gspec, ethash.NewFaker(), 1, concreteRegistry, func(ii int, block *core.BlockGen) {
-		for _, id := range [][]byte{setupId, method.ID} {
-			tx := types.NewTransaction(block.TxNonce(senderAddress), contractAddress, common.Big0, gasLimit, block.BaseFee(), id)
-			signed, err := types.SignTx(tx, signer, key)
-			if err != nil {
-				panic(err)
+		_, _, receipts := core.GenerateChainWithGenesisWithConcrete(gspec, ethash.NewFaker(), 1, concreteRegistry, func(ii int, block *core.BlockGen) {
+			for _, id := range [][]byte{setupId, method.ID} {
+				tx := types.NewTransaction(block.TxNonce(senderAddress), contractAddress, common.Big0, gasLimit, block.BaseFee(), id)
+				signed, err := types.SignTx(tx, signer, key)
+				if err != nil {
+					t.Fatal(err)
+				}
+				block.AddTx(signed)
 			}
-			block.AddTx(signed)
-		}
-	})
+		})
 
-	if len(receipts[0]) != 2 {
-		return 0, fmt.Errorf("expected 2 receipts, got %d", len(receipts))
-	}
-	setupReceipt := receipts[0][0]
-	testReceipt := receipts[0][1]
-	if setupReceipt.Status != types.ReceiptStatusSuccessful {
-		return 0, fmt.Errorf("setup failed")
-	}
-	if (testReceipt.Status == types.ReceiptStatusSuccessful) == shouldFail {
-		return 0, fmt.Errorf("test failed")
-	}
+		t.Run("validateReceipts", func(t *testing.T) {
+			if len(receipts[0]) != 2 {
+				t.Fatalf("expected 2 receipts, got %d", len(receipts[0]))
+			}
+			setupReceipt := receipts[0][0]
+			testReceipt := receipts[0][1]
 
-	if PrintLogs && len(testReceipt.Logs) > 0 {
-		fmt.Println("")
-		for ii, log := range testReceipt.Logs {
-			fmt.Printf("Logs[%d]\n", ii)
-			fmt.Println("Address :", log.Address)
-			if len(log.Topics) > 0 {
-				fmt.Println("Topics  :", log.Topics[0])
-				for _, topic := range log.Topics[1:] {
-					fmt.Println("         ", topic)
-				}
+			if setupReceipt.Status != types.ReceiptStatusSuccessful {
+				t.Fatalf("setup failed")
+			}
+			if (testReceipt.Status == types.ReceiptStatusSuccessful) == shouldFail {
+				t.Fatalf("test failed")
 			}
-			fmt.Println("Data    : 0x" + hex.EncodeToString(log.Data))
-		}
-		fmt.Println("")
-	}
 
-	return testReceipt.GasUsed, nil
+			if PrintLogs && len(testReceipt.Logs) > 0 {
+				t.Log("")
+				for ii, log := range testReceipt.Logs {
+					t.Logf("Logs[%d]", ii)
```

Comment:
This might work but will be hard to read as t.Log logs are prefixed by the test file path and line number. It's better to generate a string and then log it once like in the example I provided.

```golang
if PrintLogs && len(testReceipt.Logs) > 0 {
	for ii, log := range testReceipt.Logs {
		logStr := fmt.Sprintf("\nLogs[%d]\nAddress : %s\n", ii, log.Address)
		if len(log.Topics) > 0 {
			logStr += fmt.Sprintf("Topics  : %s\n", log.Topics[0].String())
			for _, topic := range log.Topics[1:] {
				logStr += fmt.Sprintf("         : %s\n", topic.String())
			}
		}
		logStr += fmt.Sprintf("Data    : 0x%x\n", log.Data)
		t.Log(logStr)
	}
}
```